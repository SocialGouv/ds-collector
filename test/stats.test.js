process.env.DS_API_URL = "http://fakeApi.com";
process.env.DS_ID_PROCEDURE = "4212";

const { insert, find, clear } = require("../src/db");
const { getStats, aggregate } = require("../src/stats");

const SAMPLE_DOSSIERS = [
  {
    dossier: {
      id: 1,
      created_at: "2018-05-17T12:00:21.005Z",
      updated_at: "2018-06-09T13:43:59.462Z",
      processed_at: "2008-09-01T10:05:00.000Z",
      initiated_at: "2008-05-01T10:05:00.000Z",
      state: "closed",
      champs: [
        {
          value: "hello",
          type_de_champ: {
            libelle: "answer"
          }
        }
      ]
    }
  },
  {
    dossier: {
      id: 2,
      created_at: "2018-02-12T12:00:21.005Z",
      updated_at: "2018-03-09T13:43:59.462Z",
      processed_at: "2008-09-01T10:05:00.000Z",
      initiated_at: "2008-05-01T10:05:00.000Z",
      state: "refused",
      champs: [
        {
          value: "hello2",
          type_de_champ: {
            libelle: "answer"
          }
        }
      ]
    }
  },
  {
    dossier: {
      id: 3,
      created_at: "2018-01-17T12:00:21.005Z",
      updated_at: "2018-02-09T13:43:59.462Z",
      processed_at: "2008-09-01T10:05:00.000Z",
      initiated_at: "2008-05-01T10:05:00.000Z",
      state: "closed",
      champs: [
        {
          value: "hello2",
          type_de_champ: {
            libelle: "answer"
          }
        }
      ]
    }
  },
  {
    dossier: {
      id: 4,
      created_at: "2018-01-17T12:00:21.005Z",
      updated_at: "2018-04-09T13:43:59.462Z",
      processed_at: "2008-09-01T10:05:00.000Z",
      initiated_at: "2008-05-01T10:05:00.000Z",
      state: "refused",
      champs: [
        {
          value: "hello",
          type_de_champ: {
            libelle: "answer"
          }
        }
      ]
    }
  },
  {
    dossier: {
      id: 5,
      created_at: "2018-05-01T12:00:21.005Z",
      updated_at: "2018-06-09T13:43:59.462Z",
      processed_at: "2008-09-01T10:05:00.000Z",
      initiated_at: "2008-05-01T10:05:00.000Z",
      state: "closed",
      champs: [
        {
          value: "hello",
          type_de_champ: {
            libelle: "answer"
          }
        }
      ]
    }
  },
  {
    dossier: {
      id: 6,
      created_at: "2018-05-02T12:00:21.005Z",
      updated_at: "2018-07-10T13:43:59.462Z",
      processed_at: "2008-09-01T10:05:00.000Z",
      initiated_at: "2008-05-01T10:05:00.000Z",
      state: "closed",
      champs: [
        {
          value: "hello",
          type_de_champ: {
            libelle: "answer"
          }
        }
      ]
    }
  },
  {
    dossier: {
      id: 7,
      created_at: "2018-02-21T12:00:21.005Z",
      updated_at: "2018-03-09T13:43:59.462Z",
      processed_at: "2008-09-01T10:05:00.000Z",
      initiated_at: "2008-05-01T10:05:00.000Z",
      state: "without_continuation",
      champs: [
        {
          value: "hello3",
          type_de_champ: {
            libelle: "answer"
          }
        }
      ]
    }
  },
  {
    dossier: {
      id: 8,
      created_at: "2018-03-27T12:00:21.005Z",
      updated_at: "2018-04-09T13:43:59.462Z",
      processed_at: "2008-09-01T10:05:00.000Z",
      initiated_at: "2008-05-01T10:05:00.000Z",
      state: "without_continuation",
      champs: [
        {
          value: "hello2",
          type_de_champ: {
            libelle: "answer"
          }
        }
      ]
    }
  },
  {
    dossier: {
      id: 9,
      created_at: "2017-01-17T12:00:21.005Z",
      updated_at: "2017-02-09T13:43:59.462Z",
      processed_at: "2008-09-01T10:05:00.000Z",
      initiated_at: "2008-05-01T10:05:00.000Z",
      state: "draft",
      champs: [
        {
          value: "hello2",
          type_de_champ: {
            libelle: "answer"
          }
        }
      ]
    }
  },
  {
    dossier: {
      id: 10,
      created_at: "2017-05-07T12:00:21.005Z",
      updated_at: "2018-07-09T13:43:59.462Z",
      processed_at: "2008-09-01T10:05:00.000Z",
      initiated_at: "2008-05-01T10:05:00.000Z",
      state: "closed",
      champs: [
        {
          value: "hello",
          type_de_champ: {
            libelle: "answer"
          }
        }
      ]
    }
  }
];
describe("stats", () => {
  beforeEach(() => {
    clear();
  });
  describe("getStats", () => {
    it("should compute correct stats", async () => {
      SAMPLE_DOSSIERS.forEach(dossier => insert(dossier));
      const docs = await find({}, { "dossier.id": 1 }); // sort results to get consistent snapshots
      const stats = getStats(docs);
      expect(stats).toMatchSnapshot();
    });
    it("should return aggregated libelle values", async () => {
      SAMPLE_DOSSIERS.forEach(dossier => insert(dossier));
      const docs = await find({}, { "dossier.id": 1 }); // sort results to get consistent snapshots
      const agg = aggregate(docs, "answer");
      expect(agg).toMatchSnapshot();
    });
  });
});
