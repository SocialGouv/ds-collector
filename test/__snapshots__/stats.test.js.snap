// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stats getStats should compute correct stats 1`] = `
Object {
  "count": 10,
  "daily": Object {
    "2017-01-17": Object {
      "count": 1,
      "duration": 0,
      "durations": Array [],
      "status": Object {
        "closed": 0,
        "draft": 1,
        "initiated": 0,
        "received": 0,
        "refused": 0,
        "without_continuation": 0,
      },
    },
    "2017-05-07": Object {
      "count": 1,
      "duration": 428,
      "durations": Array [
        428,
      ],
      "status": Object {
        "closed": 1,
        "draft": 0,
        "initiated": 0,
        "received": 0,
        "refused": 0,
        "without_continuation": 0,
      },
    },
    "2018-01-17": Object {
      "count": 2,
      "duration": 52.5,
      "durations": Array [
        23,
        82,
      ],
      "status": Object {
        "closed": 1,
        "draft": 0,
        "initiated": 0,
        "received": 0,
        "refused": 1,
        "without_continuation": 0,
      },
    },
    "2018-02-12": Object {
      "count": 1,
      "duration": 25,
      "durations": Array [
        25,
      ],
      "status": Object {
        "closed": 0,
        "draft": 0,
        "initiated": 0,
        "received": 0,
        "refused": 1,
        "without_continuation": 0,
      },
    },
    "2018-02-21": Object {
      "count": 1,
      "duration": 16,
      "durations": Array [
        16,
      ],
      "status": Object {
        "closed": 0,
        "draft": 0,
        "initiated": 0,
        "received": 0,
        "refused": 0,
        "without_continuation": 1,
      },
    },
    "2018-03-27": Object {
      "count": 1,
      "duration": 13,
      "durations": Array [
        13,
      ],
      "status": Object {
        "closed": 0,
        "draft": 0,
        "initiated": 0,
        "received": 0,
        "refused": 0,
        "without_continuation": 1,
      },
    },
    "2018-05-01": Object {
      "count": 1,
      "duration": 39,
      "durations": Array [
        39,
      ],
      "status": Object {
        "closed": 1,
        "draft": 0,
        "initiated": 0,
        "received": 0,
        "refused": 0,
        "without_continuation": 0,
      },
    },
    "2018-05-02": Object {
      "count": 1,
      "duration": 69,
      "durations": Array [
        69,
      ],
      "status": Object {
        "closed": 1,
        "draft": 0,
        "initiated": 0,
        "received": 0,
        "refused": 0,
        "without_continuation": 0,
      },
    },
    "2018-05-17": Object {
      "count": 1,
      "duration": 23,
      "durations": Array [
        23,
      ],
      "status": Object {
        "closed": 1,
        "draft": 0,
        "initiated": 0,
        "received": 0,
        "refused": 0,
        "without_continuation": 0,
      },
    },
  },
  "duration": 79.77777777777777,
  "monthly": Object {
    "2017-01": Object {
      "count": 1,
      "duration": 0,
      "durations": Array [],
      "status": Object {
        "closed": 0,
        "draft": 1,
        "initiated": 0,
        "received": 0,
        "refused": 0,
        "without_continuation": 0,
      },
    },
    "2017-05": Object {
      "count": 1,
      "duration": 428,
      "durations": Array [
        428,
      ],
      "status": Object {
        "closed": 1,
        "draft": 0,
        "initiated": 0,
        "received": 0,
        "refused": 0,
        "without_continuation": 0,
      },
    },
    "2018-01": Object {
      "count": 2,
      "duration": 52.5,
      "durations": Array [
        23,
        82,
      ],
      "status": Object {
        "closed": 1,
        "draft": 0,
        "initiated": 0,
        "received": 0,
        "refused": 1,
        "without_continuation": 0,
      },
    },
    "2018-02": Object {
      "count": 2,
      "duration": 20.5,
      "durations": Array [
        25,
        16,
      ],
      "status": Object {
        "closed": 0,
        "draft": 0,
        "initiated": 0,
        "received": 0,
        "refused": 1,
        "without_continuation": 1,
      },
    },
    "2018-03": Object {
      "count": 1,
      "duration": 13,
      "durations": Array [
        13,
      ],
      "status": Object {
        "closed": 0,
        "draft": 0,
        "initiated": 0,
        "received": 0,
        "refused": 0,
        "without_continuation": 1,
      },
    },
    "2018-05": Object {
      "count": 3,
      "duration": 43.666666666666664,
      "durations": Array [
        23,
        39,
        69,
      ],
      "status": Object {
        "closed": 3,
        "draft": 0,
        "initiated": 0,
        "received": 0,
        "refused": 0,
        "without_continuation": 0,
      },
    },
  },
  "status": Object {
    "closed": 5,
    "draft": 1,
    "initiated": 0,
    "received": 0,
    "refused": 2,
    "without_continuation": 2,
  },
}
`;

exports[`stats getStats should return aggregated libelle values 1`] = `
Object {
  "hello": 5,
  "hello2": 4,
  "hello3": 1,
}
`;
